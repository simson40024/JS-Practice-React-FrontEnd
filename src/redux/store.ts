import { configureStore } from "@reduxjs/toolkit";
import paginationReducer from "./slices/paginationSlice";
import userReducer from "./slices/userSlice";
import sortingReducer from "./slices/sortingSlice";
import currencyReducer from "./slices/currencySlice";
import productReducer from "./slices/productSlice";
import cartReducer from "./slices/cartSlice";

 const store = configureStore({
  reducer: {
    user: userReducer,
    pagination: paginationReducer,
    sorting: sortingReducer,
    currency: currencyReducer,
    product: productReducer,
    cart: cartReducer,
  },
});

export default store;

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;