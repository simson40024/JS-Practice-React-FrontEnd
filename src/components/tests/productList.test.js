import ProductList from "../shopping/productList";
import { createRoot } from "react-dom/client";
import { act } from "@testing-library/react";
import * as hooks from "../../hook";
import { Provider } from "react-redux";
import store from "../../redux/store";

const InitialState = {
  products: [],
  status: "loading",
  error: null,
};

const productState = {
  products: [
    { _id: "123", name: "first product", price: 30, imgUrl: "/img.ipg" },
    { _id: "134", name: "second product", price: 30, imgUrl: "/img.ipg" },
  ],
  status: "loaded",
  error: null,
};

const stateWithLoadingError = {
  products: [],
  status: "rejected",
  error: "Network error. Data don`t loaded",
};
let container = null;
beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  container.remove();
  container = null;
});

const mock = jest.spyOn(hooks, "useAppSelector");
//не забудь замінити це нижче

describe("ProductList", () => {
  it("should create ProductList with filled Products", () => {
    mock.mockReturnValue(productState);
    act(() => {
      createRoot(container).render(
        <Provider store={store}>
          <ProductList />
        </Provider>
      );
    });
    expect(container).toMatchSnapshot();
  });

  it("should create ProductList with empty Products", () => {
    jest.spyOn(hooks, "useAppSelector").mockReturnValue(InitialState);
    act(() => {
      createRoot(container).render(
        <Provider store={store}>
          <ProductList />
        </Provider>
      );
    });
    expect(container).toMatchSnapshot();
  });

  it("should create ProductList with empty Products, and show LOADING ERRER", () => {
    jest.spyOn(hooks, "useAppSelector").mockReturnValue(stateWithLoadingError);
    act(() => {
      createRoot(container).render(
        <Provider store={store}>
          <ProductList />
        </Provider>
      );
    });
    expect(container).toMatchSnapshot();
  });
});
