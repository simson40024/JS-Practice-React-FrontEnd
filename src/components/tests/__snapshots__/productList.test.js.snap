// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList should create ProductList with empty Products 1`] = `
<div>
  <div
    class="loading"
  >
    loading...
  </div>
  <div
    class="product-list"
  />
</div>
`;

exports[`ProductList should create ProductList with empty Products, and show LOADING ERRER 1`] = `
<div>
  <div
    class="loading"
  >
    loading error
  </div>
  <div
    class="product-list"
  />
</div>
`;

exports[`ProductList should create ProductList with filled Products 1`] = `
<div>
  <div
    class="product-list"
  >
    <div
      class="product"
      id="123"
    >
      <img
        alt="product in the shop"
        src="/img.ipg"
        width="250px"
      />
      <p
        class="product-name"
      >
        first product
      </p>
      <p
        class="product-price"
      >
        NaN
         
      </p>
      <div
        class="product-quantity"
      >
        <button
          role="button"
        >
          -
        </button>
        <input
          readonly=""
          type="text"
          value="1"
        />
        <button
          role="button"
        >
          +
        </button>
      </div>
      <button
        class="add-to-cart"
        role="button"
      >
        add to cart
      </button>
    </div>
    <div
      class="product"
      id="134"
    >
      <img
        alt="product in the shop"
        src="/img.ipg"
        width="250px"
      />
      <p
        class="product-name"
      >
        second product
      </p>
      <p
        class="product-price"
      >
        NaN
         
      </p>
      <div
        class="product-quantity"
      >
        <button
          role="button"
        >
          -
        </button>
        <input
          readonly=""
          type="text"
          value="1"
        />
        <button
          role="button"
        >
          +
        </button>
      </div>
      <button
        class="add-to-cart"
        role="button"
      >
        add to cart
      </button>
    </div>
  </div>
</div>
`;
